r/src/array.cpp

```diff
@@ -210,6 +210,20 @@ std::shared_ptr<Array> MakeStringArray(StringVector_ vec) {
   return MakeArray(data);
 }
 
+std::shared_ptr<Array> MakeFactorArray(Rcpp::IntegerVector_ factor) {
+  auto dict_values = MakeStringArray(Rf_getAttrib(factor, R_LevelsSymbol));
+  auto dict_type = dictionary(int32(), dict_values, Rf_inherits(factor, "ordered"));
+
+  auto n = Rf_length(factor);
+  IntegerVector r_indices(n);
+  for (R_xlen_t i = 0; i < n; i++) {
+    r_indices[i] = factor[i] == NA_INTEGER ? NA_INTEGER : (factor[i] - 1);
+  }
+  auto indices = SimpleArray<INTSXP, arrow::Int32Type>(r_indices);
```

Comment:
Does this allocate additional memory?

Reply:
No because it uses an RBuffer which borrows the memory of the R object.